src/vs/workbench/contrib/chat/common/chatModel.ts

```diff
@@ -81,6 +81,53 @@ export class ChatRequestModel implements IChatRequestModel {
 	}
 }
 
+
+interface ResponsePart { string: IMarkdownString; resolving?: boolean }
+class Response {
+	private _responseParts: ResponsePart[];
+	private _responseRepr: IMarkdownString;
+
+	get value(): IMarkdownString {
+		return this._responseRepr;
+	}
+
+	constructor(value: IMarkdownString) {
+		this._responseRepr = value;
+		this._responseParts = [{ string: value }];
+	}
+
+	updateContent(responsePart: string | { placeholder: string; resolvedContent?: Promise<string> }): void {
+		if (typeof responsePart === 'string') {
+			const responsePartLength = this._responseParts.length - 1;
+			const lastResponsePart = this._responseParts[responsePartLength];
+
+			if (lastResponsePart.resolving === true) {
+				// The last part is resolving, start a new part
+				this._responseParts.push({ string: new MarkdownString(responsePart) });
+			} else {
+				// Combine this part with the last, non-resolving part
+				this._responseParts[responsePartLength] = { string: new MarkdownString(lastResponsePart.string.value + responsePart) };
+			}
+
+			this._updateRepr();
+		} else {
+			// Add a new resolving part
+			const responsePosition = this._responseParts.push({ string: new MarkdownString(responsePart.placeholder), resolving: true });
+			this._updateRepr();
+
+			responsePart.resolvedContent?.then((content) => {
```

Comment:
Why is `resolvedContent` optional, don't you need the promise to be able to say when it's done?